### ISSUE TITLE
🐛[Bug] Support for `@shopify/app-bridge-react` V4

### ISSUE THREAD

**Issue by SiddharthSham** [Shopify/discount-app-components#196]
## Issue summary

After upgrading to V4 of `@shopify/app-bridge-react` as per the [migration docs](https://shopify.dev/docs/api/app-bridge/migration-guide), this library throws an error due to trying to import the now deprecated `<Provider />` component in [this test](https://github.com/Shopify/discount-app-components/blob/main/src/components/DiscountAppComponentsTestProvider/DiscountAppComponentsTestProvider.tsx) file.

## Expected behavior
No error should be thrown.

This library should also be updated to use the latest version of `@shopify/app-bridge-react`

## Actual behavior

The thrown error is:
```
[ERROR] No matching export in 
"node_modules/.pnpm/@shopify+app-bridge-react@4.0.0_react-dom@18.2.0_react@18.2.0/node_modules/@shopify/app-bridge-react/dist/index.js" for import "Provider"

node_modules/.pnpm/@shopify+discount-app-components@2.1.2_@shopify+app-bridge-react@4.0.0_@shopify+app-bridge@3._fdrtohdej5n7wsk37q6fjgkki4/node_
modules/@shopify/discount-app-components/build/esm/components/DiscountAppComponentsTestProvider/DiscountAppComponentsTestProvider.js:3:9:
3 │ import { Provider } from '@shopify/app-bridge-react';
```

## Steps to reproduce the problem

1. Start from the Remix app template, add the `discount-app-components` package
2. Update `@shopify/app-bridge-react` to the latest version
3. Try to run the app

## Specifications

- Are you using the React components? (Y/N): Y
- discount-app-components version number: 2.1.2
- Browser: N/A
- Device: N/A
- Operating System: N/A

**Comment by Yamashou** [comment]
Could you provide an update on this issue? I would appreciate it if you could update it if possible.

**Comment by echocrow** [comment]
Running into the same issue too, which is preventing us from building a critical discount bundle app. Following the latest official guide to a tee does not help either.

Tried downgrading to V3, but that either fails with
```
[vite] Named export 'Provider' not found. The requested module '@shopify/app-bridge-react' is a CommonJS module, which may not support all module.exports as named exports.
```
or, when following the default export suggestion, does compile, but then clicks and other events stop working entirely.

**Comment by paulomenezes** [comment]
I created a fork of this repository to fix the issue and successfully built my remix app using `@shopify/app-bridge-react V4` https://github.com/Alpha-Omega-Agency/discount-app-components. 
I didn't open the PR because there is an error with `const redirect = Redirect.create(app);` that I'm not sure how to fix, so I just ignored it because it isn't impacting my code.

**Comment by ijohns1904** [comment]
@devisscher Can you share any updates on this issue? Thanks!

**Comment by philkendallsimba** [comment]
We are facing this issue too. An update would be good please. Thanks.

**Comment by pnmcosta** [comment]
> I didn't open the PR because there is an error with `const redirect = Redirect.create(app);` that I'm not sure how to fix, so I just ignored it because it isn't impacting my code.

@paulomenezes I believe all uses of `redirect` can be replaced with `open('/settings', '_self');` see https://shopify.dev/docs/api/app-bridge-library/apis/navigation tap Javascript tab for each example.

Have you come across any other issues using your fork? I haven't reviewed your fork yet, but am concerned that this packages use of localization maybe affected.

**Comment by FancyMyReality** [comment]
Hi @devisscher, is there a status update on this bug? Waiting on resolution and cannot continue our development.

**Comment by RivermintGlenn** [comment]
> Running into the same issue too, which is preventing us from building a critical discount bundle app. Following the latest official guide to a tee does not help either.
> 
> Tried downgrading to V3, but that either fails with
> 
> ```
> [vite] Named export 'Provider' not found. The requested module '@shopify/app-bridge-react' is a CommonJS module, which may not support all module.exports as named exports.
> ```
> 
> or, when following the default export suggestion, does compile, but then clicks and other events stop working entirely.

@echocrow 

I was having the same issue, whereby even after downgrading, the commonJs error was being throw.

It turns out the issue is with the new vite + remix integration. 

I was able to fix the issue by following [this guide](https://remix.run/docs/en/dev/future/vite#esm--cjs), specifically by using the [vite-plugin-cjs-interop plugin](https://github.com/cyco130/vite-plugin-cjs-interop) plugin.

```js
// vite.config.js
cjsInterop({
      // List of CJS dependencies that require interop
      dependencies: ["@shopify/app-bridge-react"],
}),
```

**Comment by polymorphtech** [comment]
The CJS Interop plugin removes the vite error. But in the browser console, we now observe

Uncaught SyntaxError: The requested module '/node_modules/@shopify/polaris/build/esm/styles.css?t=1713557476064' does not provide an export named 'default' (at app.tsx:4:1)

**Comment by RivermintGlenn** [comment]
@polymorphtech 

I believe this is an issue with vite remix.

You can import the `polarisStyles` directly into your base component and can remove the link export.

_(Note: I am not the best to listen to here and would love someone to add more technical context. I'm 🤠ing it)_

```jsx
import "@shopify/polaris/build/esm/styles.css";

// export const links = () => [{ rel: "stylesheet", href: polarisStyles }];
```

**Comment by BornAgainHooligan** [comment]
Hey team/ @devisscher!

Just checking in for an update on this as it has been three weeks without an update. Can we please get an outline of where we are at with this?

**Comment by Yamashou** [comment]
@devisscher @mathiusj 
I hope this issue will be resolved and released soon. We are looking forward to the update of app-bridge as soon as possible.

**Comment by philkendallsimba** [comment]
I used @paulomenezes fork of the repo in our app, which is being used in production, instead of the default one, and am having good success with.
I am still hopeful we will get an official update soon though, as I don't like being dependent on the fork.

**Comment by Michael-Gibbons** [comment]
@devisscher Why did you close this issue? Doesn't this completely break development? My Coworker and I just spent around an hour trying to generate a new app extension following this documentation:

https://shopify.dev/docs/apps/build/discounts/experience/build-discounts-function

On both Mac and Windows we were both getting a dependency tree error relating to app bridge v4, since the dependency lives within the shopify package code we can't downgrade, making the creation of new extension apps impossible.

**Comment by echocrow** [comment]
as far as I can tell, the interop bug that prevented bundling with App Bridge v4 was fixed in #203.

after that, we were able to build an embedded discount admin UI using this project in combination with App Bridge v4.

that said, IIRC the [tutorial](https://shopify.dev/docs/apps/build/discounts/experience/build-ui#step-1-set-up-discounts-dependencies) is outdated (e.g. `Provider` from `"@shopify/app-bridge-react"` is an artifact from v3 and need not be imported in v4), and `app-bridge.js` from the [migration docs](https://shopify.dev/docs/api/app-bridge/migration-guide) isn't always needed.

we started off using Shopify's Remix template for embedded apps, and loosely followed the discount tutorials from there, referencing the App Bridge v4 migration in the process. a little tedious, but worked in the end.
